### PR TITLE
fix(1010): [1][medium] add chainPR annotation.

### PR DESCRIPTION
**Pull request by wdstar**
## Context

prChain flag is moved from top-level setting to pipeline-level annotation.
- Current
```
prChain: true
jobs:
   main:
       ....
```
- After
```
annotations:
  screwdriver.cd/chainPR: true 
jobs:
   main:
       ....
```

## Objective

This PR adds value resolution of the `screwdriver.cd/chainPR` annotation and the plugin-level setting, which is added for maintenance of similarity with `restrictPR`.

## References

- https://github.com/screwdriver-cd/screwdriver/issues/1010#issuecomment-466188824

### CODE REVIEW

File: plugins/webhooks/index.js [hunk 1]
```diff
@@ -207,7 +211,8 @@ async function createPREvents(options, request) {
             configPipelineSha,
             startFrom: `~pr:${branch}`,
             changedFiles,
-            causeMessage: `${action} by ${userDisplayName}`
+            causeMessage: `${action} by ${userDisplayName}`,
+            chainPR: resolvedChainPR
```
Comment (wdstar) [inline plugins/webhooks/index.js]: This `eventConfig.chainPR` flag is finally passed to `eventFactory.create()`.

**Comment by wdstar** [comment]
The test about addition of `~release` and `~tag` nodes to workflowGraph fails.
https://cd.screwdriver.cd/pipelines/1/builds/98121

File: plugins/webhooks/index.js [hunk 2]
```diff
@@ -617,7 +645,8 @@ exports.register = (server, options, next) => {
     const pluginOptions = joi.attempt(options, joi.object().keys({
         username: joi.string().required(),
         ignoreCommitsBy: joi.array().items(joi.string()).optional(),
-        restrictPR: joi.string().valid('all', 'none', 'branch', 'fork').optional()
+        restrictPR: joi.string().valid('all', 'none', 'branch', 'fork').optional(),
+        chainPR: joi.boolean().optional()
```
Comment (sakka2) [inline plugins/webhooks/index.js]: should add param `chainPR` to doc comment.

File: plugins/webhooks/index.js [hunk 3]
```diff
@@ -232,6 +237,21 @@ async function createPREvents(options, request) {
...
+ * @param  {Boolean}  chainPR              Plugin Chain PR flag
+ * @param  {Pipeline} pipeline             Pipeline
+ * @param  {Object}   pipeline.annotations Pipeline-level annotations
+ * @return {Boolean}
+ */
+function resolveChainPR(chainPR, pipeline) {
+    const defaultChainPR = typeof chainPR === 'undefined' ? false : chainPR;
+    const annotChainPR = pipeline.annotations[ANNOT_CHAIN_PR];
+
+    return typeof annotChainPR === 'undefined' ? defaultChainPR : annotChainPR;
```
Comment (sakka2) [inline plugins/webhooks/index.js]: [ask] Is there any case that value of chainPR becomes `undefined`?

File: plugins/webhooks/index.js (hunk 2 above)
Comment (wdstar) [reply plugins/webhooks/index.js]: I added it. Thanks.

File: plugins/webhooks/index.js (hunk 3 above)
Comment (wdstar) [reply plugins/webhooks/index.js]: Yes, when there is no `screwdriver.cd/chainPR` annotation in a `screwdriver.yaml`.

File: plugins/webhooks/index.js [hunk 4]
```diff
@@ -232,6 +237,21 @@ async function createPREvents(options, request) {
...
+/**
+ * Resolve ChainPR flag
+ * @method resolveChainPR
+ * @param  {Boolean}  chainPR              Plugin Chain PR flag
+ * @param  {Pipeline} pipeline             Pipeline
+ * @param  {Object}   pipeline.annotations Pipeline-level annotations
+ * @return {Boolean}
+ */
+function resolveChainPR(chainPR, pipeline) {
+    const defaultChainPR = typeof chainPR === 'undefined' ? false : chainPR;
```
Comment (tkyi) [inline plugins/webhooks/index.js]: does this work too?
```suggestion
    const defaultChainPR = chainPR || false;
```

File: plugins/webhooks/index.js (hunk 3 above)
Comment (tkyi) [inline plugins/webhooks/index.js]: Does this work here?
```suggestion
    return annotChainPR || defaultChainPR;
```

Comment (wdstar) [reply plugins/webhooks/index.js]: No. If `annotChainPR` is false and `defaultChainPR` is true, it returns true, but we expect false.

File: plugins/webhooks/index.js (hunk 4 above)
Comment (wdstar) [reply plugins/webhooks/index.js]: Yes, but we should not use `||` short circuit evaluation about **Boolean** type.
Please refer to the following comments too.

File: plugins/webhooks/index.js (hunk 3 above)
Comment (tkyi) [reply plugins/webhooks/index.js]: Ah thanks for explaining, makes sense

**Pull request merged**